desk/app/channels-server.hoon

```diff
@@ -842,4 +1004,287 @@
+    =.  config.hook  (~(put by config.hook) nest config)
+    =.  cor  (give-hook-response [%config id nest config])
+    ho-core
+  ++  ho-wait
+    |=  [=origin:h schedule=$@(@dr schedule:h) =config:h]
+    ^+  ho-core
+    =/  schedule
+      ?:  ?=(@ schedule)  [now.bowl schedule]
+      schedule
+    =/  crons  (~(gut by crons.hooks) id *(map origin:h cron:h))
+    =/  =cron:h  [id schedule config]
+    =.  crons.hooks
+      =-  (~(put by crons.hooks) id.hook -)
+      (~(put by crons) origin cron)
+    =.  cor  (schedule-cron origin cron)
+    =.  cor  (give-hook-response [%wait id origin schedule config])
+    ho-core
+  ++  ho-rest
+    |=  =origin:h
+    ^+  ho-core
+    =/  crons  (~(got by crons.hooks) id)
+    =/  cron  (~(got by crons) origin)
+    =.  crons.hooks
+      (~(put by crons.hooks) id (~(del by crons) origin))
+    =.  cor  (unschedule-cron origin cron)
+    =.  cor  (give-hook-response [%rest id origin])
+    ho-core
+  ++  ho-run-single
+    |=  [=event:h prefix=tape =origin:h =config:h]
+    =/  channel
+      ?@  origin  ~
+      ?~  ch=(~(get by v-channels) origin)  ~
+      `[origin u.ch]
+    =/  =context:h  (get-hook-context channel config)
+    =/  return=(unit return:h)
+      (run-hook:utils [event context(hook hook)] hook)
+    ?~  return
+      %-  (slog (crip "{prefix} {<id>} failed, running on {<origin>}") ~)
+      ho-core
+    %-  (slog (crip "{prefix} {<id>} ran on {<origin>}") ~)
+    =.  hook  hook(state new-state.u.return)
+    =.  cor  (run-hook-effects effects.u.return origin)
+    ho-core
+  --
+++  run-hooks
+  |=  [=event:h =nest:c default=cord]
+  ^-  [(each event:h tang) _cor]
+  =;  [result=(each event:h tang) effects=(list effect:h)]
+    [result (run-hook-effects effects nest)]
+  =/  current-event  event
+  =|  effects=(list effect:h)
+  =/  order  (~(got by order.hooks) nest)
+  =/  channel  `[nest (~(got by v-channels) nest)]
+  =/  =context:h  (get-hook-context channel *config:h)
+  |-
+  ?~  order
+    [&+current-event effects]
+  =*  next  $(order t.order)
+  =/  hook  (~(got by hooks.hooks) i.order)
+  =/  ctx  context(hook hook, config (~(gut by config.hook) nest ~))
+  =/  return=(unit return:h)
+    (run-hook:utils [current-event ctx] hook)
+  ?~  return  next
+  =*  result  result.u.return
+  =.  effects  (weld effects effects.u.return)
+  =.  hooks.hooks  (~(put by hooks.hooks) i.order hook(state new-state.u.return))
+  ?:  ?=(%denied -.result)
+    [|+~[(fall msg.result default)] effects]
+  =.  current-event  new.result
+  next
+++  wakeup-hook
+  |=  =(pole knot)
+  ^+  cor
+  ?+  pole  ~|(bad-arvo-take/pole !!)
+      [%delayed id=@ ~]
+    =/  =id:h  (slav %uv id.pole)
+    ?~  delay=(~(get by delayed.hooks) id)  cor
+    ::  make sure we clean up
+    =.  delayed.hooks  (~(del by delayed.hooks) id)
+    ::  ignore premature fires
+    ?:  (lth now.bowl fires-at.u.delay)  cor
+    =*  origin  origin.u.delay
+    =/  hook  (~(got by hooks.hooks) hook.u.delay)
```

Comment:
Another double plural, more understandable than with `cron`, but maybe we could find a new name for either.